In [1]:
import gzip
import shutil
import os
import pandas as pd
from ast import literal_eval
import json
from datetime import datetime
from sqlalchemy import create_engine


In [2]:
import os

# Get the current directory
current_directory = os.getcwd()
print("Current Directory:", current_directory)

Current Directory: /content


Getting data from json files into a dataframe for analysis

In [5]:
for i in os.listdir():
  if 'json' in i:
    with gzip.open(i, 'rb') as f_in:
          with open(i.replace('.gz',''), 'wb') as f_out:
            shutil.copyfileobj(f_in, f_out)

In [32]:

receipts = pd.read_json('receipts.json',lines=True)
brands=pd.read_json('brands.json',lines=True)
users=pd.read_json('users.json',lines=True)

In [33]:
receipts.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

Splitting up the rewardsreceiptItemList column into multiple columns and replacing nulls with Nan

In [8]:
receipts = receipts.explode('rewardsReceiptItemList')
receipts.reset_index(inplace=True)

In [9]:

receipts = receipts.fillna({'rewardsReceiptItemList':'{}'})
receipts['rewardsReceiptItemList'] = receipts['rewardsReceiptItemList'].apply(lambda x:str(x))
receipts['rewardsReceiptItemList'] = receipts['rewardsReceiptItemList'].apply(literal_eval)


In [10]:
receipts.head()

index                                   _id  bonusPointsEarned  \
0      0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1      1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2      1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
3      2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
4      3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2  Receipt number 5 completed, bonus point schedu...   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687483000}  {'$date': 1609687483000}   
3  {'$date': 1609687537000}  {'$date': 1609687537000}   
4  {'$date': 1609687534000}  {'$date': 1609687534000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2  {'$date': 1609687483000}  {'$date': 1609687488000}   
3                       NaN  {'$date': 1609687542000}   
4  {'$date': 1609687534000}  {'$date': 1609687539000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
3                       NaN           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  {'barcode': '4011', 'description': 'ITEM NOT F...   
1                 2.0  {'barcode': '4011', 'description': 'ITEM NOT F...   
2                 2.0  {'barcode': '028400642255', 'description': 'DO...   
3                 1.0  {'needsFetchReview': False, 'partnerItemId': '...   
4                 4.0  {'barcode': '4011', 'description': 'ITEM NOT F...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
3             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
4             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6

In [11]:
receipts_split = pd.json_normalize(receipts['rewardsReceiptItemList'],errors='ignore')

In [12]:
receipts_split=receipts_split.add_prefix('rewardsReceiptItemList_')

In [13]:
receipts_split.head()

rewardsReceiptItemList_barcode  \
0                           4011   
1                           4011   
2                   028400642255   
3                            NaN   
4                           4011   

                  rewardsReceiptItemList_description  \
0                                     ITEM NOT FOUND   
1                                     ITEM NOT FOUND   
2  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...   
3                                                NaN   
4                                     ITEM NOT FOUND   

  rewardsReceiptItemList_finalPrice rewardsReceiptItemList_itemPrice  \
0                             26.00                            26.00   
1                                 1                                1   
2                             10.00                            10.00   
3                               NaN                              NaN   
4                             28.00                            28.00   

  rewardsReceiptItemList_needsFetchReview  \
0                                   False   
1                                     NaN   
2                                    True   
3                                   False   
4                                   False   

  rewardsReceiptItemList_partnerItemId  \
0                                    1   
1                                    1   
2                                    2   
3                                    1   
4                                    1   

  rewardsReceiptItemList_preventTargetGapPoints  \
0                                          True   
1                                           NaN   
2                                          True   
3                                          True   
4                                          True   

   rewardsReceiptItemList_quantityPurchased  \
0                                       5.0   
1                                       1.0   
2                                       1.0   
3                                       NaN   
4                                       4.0   

  rewardsReceiptItemList_userFlaggedBarcode  \
0                                      4011   
1                                       NaN   
2                              028400642255   
3                                      4011   
4                                      4011   

  rewardsReceiptItemList_userFlaggedNewItem  ...  \
0                                      True  ...   
1                                       NaN  ...   
2                                      True  ...   
3                                      True  ...   
4                                      True  ...   

  rewardsReceiptItemList_itemNumber  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   

   rewardsReceiptItemList_originalMetaBriteQuantityPurchased  \
0                                                NaN           
1                                                NaN           
2                                                NaN           
3                                                NaN           
4                                                NaN           

  rewardsReceiptItemList_pointsEarned rewardsReceiptItemList_targetPrice  \
0                                 NaN                                NaN   
1                                 NaN                                NaN   
2                                 NaN                                NaN   
3                                 NaN                                NaN   
4                                 NaN                                NaN   

  rewardsReceiptItemList_competitiveProduct  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       

Merging back the split columns and the actual columns to a new dataset

In [14]:
receipts1 = pd.merge(receipts, receipts_split, left_index=True, right_index=True, how='outer')

In [15]:
receipts1.head()

index                                   _id  bonusPointsEarned  \
0      0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1      1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2      1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
3      2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
4      3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2  Receipt number 5 completed, bonus point schedu...   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687483000}  {'$date': 1609687483000}   
3  {'$date': 1609687537000}  {'$date': 1609687537000}   
4  {'$date': 1609687534000}  {'$date': 1609687534000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2  {'$date': 1609687483000}  {'$date': 1609687488000}   
3                       NaN  {'$date': 1609687542000}   
4  {'$date': 1609687534000}  {'$date': 1609687539000}   

          pointsAwardedDate  pointsEarned  ...  \
0  {'$date': 1609687531000}         500.0  ...   
1  {'$date': 1609687483000}         150.0  ...   
2  {'$date': 1609687483000}         150.0  ...   
3                       NaN           5.0  ...   
4  {'$date': 1609687534000}           5.0  ...   

  rewardsReceiptItemList_itemNumber  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   

   rewardsReceiptItemList_originalMetaBriteQuantityPurchased  \
0                                                NaN           
1                                                NaN           
2                                                NaN           
3                                                NaN           
4                                                NaN           

  rewardsReceiptItemList_pointsEarned rewardsReceiptItemList_targetPrice  \
0                                 NaN                                NaN   
1                                 NaN                                NaN   
2                                 NaN                                NaN   
3                                 NaN                                NaN   
4                                 NaN                                NaN   

   rewardsReceiptItemList_competitiveProduct  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  rewardsReceiptItemList_originalFinalPrice  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   

  rewardsReceiptItemList_originalMetaBriteItemPrice  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  rewardsReceiptItemList_deleted rewardsReceiptItemList_priceAfterCoupon  \
0                            NaN                                     NaN   
1                            NaN                                     NaN   
2                            NaN                        

Doing the same analysis for cpg column in brands

In [16]:
brands_split = pd.json_normalize(brands['cpg'])
brands_split = brands_split.add_prefix('cpg_')

In [17]:

brands1 = pd.merge(brands, brands_split, left_index=True, right_index=True, how='outer')

In [18]:
brands1.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  \
0  test brand @1612366101024       0.0                            NaN   
1                  Starbucks       0.0                      STARBUCKS   
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176   
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051   
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827   

  cpg_$ref              cpg_$id.$oid  
0     Cogs  601ac114be37ce2ead437550  
1     Cogs  5332f5fbe4b03c9a25efd0ba  
2     Cogs  601ac142be37ce2ead437559  
3     Cogs  601ac142be37ce2ead437559  
4     Cogs  5332fa12e4b03c9a25efd1e7

Getting the values of the dictionaries in receipts table

In [19]:

receipts1['_id'] = receipts1['_id'].apply(lambda x: x['$oid'])

Converting The time which is non formatted which was in milliseconds into date time format

In [20]:
def date_converter(x):
  try:
    return(datetime.utcfromtimestamp(int(x['$date'])/1000).strftime('%Y-%m-%d %H:%M:%S'))
  except TypeError:
    return(None)

In [21]:
receipts1['createDate'] = receipts1['createDate'].apply(lambda x: date_converter(x))
receipts1['dateScanned'] = receipts1['dateScanned'].apply(lambda x: date_converter(x))
receipts1['finishedDate'] = receipts1['finishedDate'].apply(lambda x: date_converter(x))
receipts1['modifyDate'] = receipts1['modifyDate'].apply(lambda x: date_converter(x))
receipts1['pointsAwardedDate'] = receipts1['pointsAwardedDate'].apply(lambda x: date_converter(x))
receipts1['purchaseDate'] = receipts1['purchaseDate'].apply(lambda x: date_converter(x))

In [22]:
receipts1.head()

index                       _id  bonusPointsEarned  \
0      0  5ff1e1eb0a720f0523000575              500.0   
1      1  5ff1e1bb0a720f052300056b              150.0   
2      1  5ff1e1bb0a720f052300056b              150.0   
3      2  5ff1e1f10a720f052300057a                5.0   
4      3  5ff1e1ee0a7214ada100056f                5.0   

                             bonusPointsEarnedReason           createDate  \
0  Receipt number 2 completed, bonus point schedu...  2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
2  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
3                         All-receipts receipt bonus  2021-01-03 15:25:37   
4                         All-receipts receipt bonus  2021-01-03 15:25:34   

           dateScanned         finishedDate           modifyDate  \
0  2021-01-03 15:25:31  2021-01-03 15:25:31  2021-01-03 15:25:36   
1  2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
2  2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
3  2021-01-03 15:25:37                 None  2021-01-03 15:25:42   
4  2021-01-03 15:25:34  2021-01-03 15:25:34  2021-01-03 15:25:39   

     pointsAwardedDate  pointsEarned  ... rewardsReceiptItemList_itemNumber  \
0  2021-01-03 15:25:31         500.0  ...                               NaN   
1  2021-01-03 15:24:43         150.0  ...                               NaN   
2  2021-01-03 15:24:43         150.0  ...                               NaN   
3                 None           5.0  ...                               NaN   
4  2021-01-03 15:25:34           5.0  ...                               NaN   

   rewardsReceiptItemList_originalMetaBriteQuantityPurchased  \
0                                                NaN           
1                                                NaN           
2                                                NaN           
3                                                NaN           
4                                                NaN           

  rewardsReceiptItemList_pointsEarned rewardsReceiptItemList_targetPrice  \
0                                 NaN                                NaN   
1                                 NaN                                NaN   
2                                 NaN                                NaN   
3                                 NaN                                NaN   
4                                 NaN                                NaN   

   rewardsReceiptItemList_competitiveProduct  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  rewardsReceiptItemList_originalFinalPrice  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   

  rewardsReceiptItemList_originalMetaBriteItemPrice  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  rewardsReceiptItemList_deleted rewardsReceiptItemList_priceAfterCoupon  \
0                            NaN                                     NaN   
1                            NaN                                     NaN   
2                            NaN                                     NaN   
3                            NaN                                     NaN   
4                            NaN                                     NaN   

  rewardsReceiptItemList_metabriteCampaignId  
0                                        NaN  
1                

Doing the same cleaning steps for users table

In [34]:
users['_id'] = users['_id'].apply(lambda x: x['$oid'])
users['createdDate'] = users['createdDate'].apply(lambda x: date_converter(x))
users['lastLogin'] = users['lastLogin'].apply(lambda x: date_converter(x))

In [35]:
users

_id  active          createdDate  \
0    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
1    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
2    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
3    5ff1e1eacfcf6c399c274ae6    True  2021-01-03 15:25:30   
4    5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
..                        ...     ...                  ...   
490  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
491  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
492  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
493  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
494  54943462e4b07e684157a532    True  2014-12-19 14:21:22   

               lastLogin         role signUpSource state  
0    2021-01-03 15:25:37     consumer        Email    WI  
1    2021-01-03 15:25:37     consumer        Email    WI  
2    2021-01-03 15:25:37     consumer        Email    WI  
3    2021-01-03 15:25:30     consumer        Email    WI  
4    2021-01-03 15:25:37     consumer        Email    WI  
..                   ...          ...          ...   ...  
490  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
491  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
492  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
493  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
494  2021-03-05 16:52:23  fetch-staff          NaN   NaN  

[495 rows x 7 columns]

Doing the same analysis for brands table

In [25]:

brands1['_id'] = brands1['_id'].apply(lambda x: x['$oid'])

In [26]:
brands1.head()

_id       barcode        category      categoryCode  \
0  601ac115be37ce2ead437551  511111019862          Baking            BAKING   
1  601c5460be37ce2ead43755f  511111519928       Beverages         BEVERAGES   
2  601ac142be37ce2ead43755d  511111819905          Baking            BAKING   
3  601ac142be37ce2ead43755a  511111519874          Baking            BAKING   
4  601ac142be37ce2ead43755e  511111319917  Candy & Sweets  CANDY_AND_SWEETS   

                                                 cpg  \
0  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  \
0  test brand @1612366101024       0.0                            NaN   
1                  Starbucks       0.0                      STARBUCKS   
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176   
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051   
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827   

  cpg_$ref              cpg_$id.$oid  
0     Cogs  601ac114be37ce2ead437550  
1     Cogs  5332f5fbe4b03c9a25efd0ba  
2     Cogs  601ac142be37ce2ead437559  
3     Cogs  601ac142be37ce2ead437559  
4     Cogs  5332fa12e4b03c9a25efd1e7

In [25]:
receipts1.drop(columns=['rewardsReceiptItemList'],axis=1)


index                       _id  bonusPointsEarned  \
0         0  5ff1e1eb0a720f0523000575              500.0   
1         1  5ff1e1bb0a720f052300056b              150.0   
2         1  5ff1e1bb0a720f052300056b              150.0   
3         2  5ff1e1f10a720f052300057a                5.0   
4         3  5ff1e1ee0a7214ada100056f                5.0   
...     ...                       ...                ...   
7376   1115  603d0b710a720fde1000042a                NaN   
7377   1116  603cf5290a720fde10000413                NaN   
7378   1117  603ce7100a7217c72c000405               25.0   
7379   1117  603ce7100a7217c72c000405               25.0   
7380   1118  603c4fea0a7217c72c000389                NaN   

                                bonusPointsEarnedReason           createDate  \
0     Receipt number 2 completed, bonus point schedu...  2021-01-03 15:25:31   
1     Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
2     Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
3                            All-receipts receipt bonus  2021-01-03 15:25:37   
4                            All-receipts receipt bonus  2021-01-03 15:25:34   
...                                                 ...                  ...   
7376                                                NaN  2021-03-01 15:42:41   
7377                                                NaN  2021-03-01 14:07:37   
7378                        COMPLETE_NONPARTNER_RECEIPT  2021-03-01 13:07:28   
7379                        COMPLETE_NONPARTNER_RECEIPT  2021-03-01 13:07:28   
7380                                                NaN  2021-03-01 02:22:34   

              dateScanned         finishedDate           modifyDate  \
0     2021-01-03 15:25:31  2021-01-03 15:25:31  2021-01-03 15:25:36   
1     2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
2     2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
3     2021-01-03 15:25:37                 None  2021-01-03 15:25:42   
4     2021-01-03 15:25:34  2021-01-03 15:25:34  2021-01-03 15:25:39   
...                   ...                  ...                  ...   
7376  2021-03-01 15:42:41                 None  2021-03-01 15:42:41   
7377  2021-03-01 14:07:37                 None  2021-03-01 14:07:37   
7378  2021-03-01 13:07:28                 None  2021-03-01 13:07:29   
7379  2021-03-01 13:07:28                 None  2021-03-01 13:07:29   
7380  2021-03-01 02:22:34                 None  2021-03-01 02:22:34   

        pointsAwardedDate  pointsEarned  ...  \
0     2021-01-03 15:25:31         500.0  ...   
1     2021-01-03 15:24:43         150.0  ...   
2     2021-01-03 15:24:43         150.0  ...   
3                    None           5.0  ...   
4     2021-01-03 15:25:34           5.0  ...   
...                   ...           ...  ...   
7376                 None           NaN  ...   
7377                 None           NaN  ...   
7378                 None          25.0  ...   
7379                 None          25.0  ...   
7380                 None           NaN  ...   

     rewardsReceiptItemList_itemNumber  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   
...                                ...   
7376                               NaN   
7377                               NaN   
7378                               NaN   
7379                               NaN   
7380                               NaN   

      rewardsReceiptItemList_originalMetaBriteQuantityPurchased  \
0                                                   NaN           
1                                                   NaN           
2                                                   NaN           
3                                                   NaN           
4                                                   Na

In [28]:
brands1.columns

Index(['_id', 'barcode', 'category', 'categoryCode', 'cpg', 'name', 'topBrand',
       'brandCode', 'cpg_$ref', 'cpg_$id.$oid'],
      dtype='object')

In [ ]:
brands1.drop(columns=['cpg'],axis=1)


In [69]:
# Convert DataFrame to Excel file
brands1.to_excel('brands.xlsx', index=False)
receipts1.to_excel('receipts.xlsx', index=False)
users.to_excel('users.xlsx', index=False)

**Third: Evaluate Data Quality Issues in the Data Provided**
We have already cleaned receipts table column (RewardreceiptList) and normalised it. We have also cleaned brands table (cpg column).

There were many columns which were in the form of dictionaries, we also brought them into tabular and relational database format for analysis.

Checking if there is any duplicate data in users table

In [36]:
user_duplicates = users.duplicated()

In [38]:
user_duplicates.sum()

283

In [42]:
user_duplicates.sum()/users.shape[0]

0.5717171717171717

1) Out of 495 columns we see that 283 are duplicates. Which is 57% of the table is duplicates

In [45]:
receipts1.shape[0]

7381

In [44]:
receipts1['pointsEarned'].isnull().sum()

1128

In [46]:
receipts1['pointsEarned'].isnull().sum()/receipts1.shape[0]

0.15282482048502913

2) About 15% of data is not populated with Points earned. Since this is a very critical column and business of Fetch depends on the marketing of brands and increasing customer base. This discrepancy is not expected.

In [47]:
receipts1['rewardsReceiptItemList_brandCode'].nunique()

227

In [48]:
brands1['brandCode'].nunique()

897

Since the joining column of receipts and brands is the brandcde. Lets check if these 227 atleast are present in the brands table

In [67]:
non_matching_values = receipts1[~receipts1['rewardsReceiptItemList_brandCode'].isin(brands1['brandCode'])]['rewardsReceiptItemList_brandCode']

In [68]:
df=pd.DataFrame(non_matching_values)

In [69]:
df['rewardsReceiptItemList_brandCode'].nunique()


186

3) Out of those 227, we see that 186 are not present in the brands table. This is a lot more bigger problem than expected as we will get skewed results.

In [73]:
users['signUpSource'].unique()

array(['Email', 'Google', nan], dtype=object)

4) Little worried about why there are only two platforms to signup

In [75]:
receipts1.isnull().mean()

index                                                        0.000000
_id                                                          0.000000
bonusPointsEarned                                            0.189812
bonusPointsEarnedReason                                      0.189812
createDate                                                   0.000000
dateScanned                                                  0.000000
finishedDate                                                 0.191167
modifyDate                                                   0.000000
pointsAwardedDate                                            0.176263
pointsEarned                                                 0.152825
purchaseDate                                                 0.062051
purchasedItemCount                                           0.065574
rewardsReceiptItemList                                       0.000000
rewardsReceiptStatus                                         0.000000
totalSpent          

5) Lot of null values in receipt items column which makes it difficult for analysis